blockchain/utxocache.go

```diff
@@ -235,6 +235,125 @@ type utxoBatcher interface {
 	getEntries(outpoints []wire.OutPoint) ([]*UtxoEntry, error)
 }
 
+// mapSlice is a slice of maps for Utxo entries.  The slice of maps are needed to
+// guarantee that the map will only take up N amount of bytes.  As of v1.20, the
+// go runtime will allocate 2^N + few extra buckets, meaning that for large N, we'll
+// allocate a lot of extra memory if the amount of entries goes over the previously
+// allocated buckets.  A slice of maps allows us to have a better control of how much
+// total memory gets allocated by all the maps.
+type mapSlice struct {
```

Comment:
Interesting...in the commit body how was the gap of choosing a size of 18,000MiB leading to an allocation of 13,000MiB identified?

Reply:
After the calculation of the entries, the go map will always allocate 2^N buckets with a few extra overflow buckets if the given number of entries are slightly above what 2^N buckets can store. The estimate for the amount of entries that can fit is 218,103,808. Each bucket for us is 368bytes so N=25 is chosen. This ends up being ~11,776MiB with overflow buckets accounting for ~2,000MiB.

When the map is full, the number is doubled, resulting in ~27,000MiB.

You can verify with the go compiler with:

```
var x map[wire.OutPoint]*UtxoEntry
var entries int = CalculateEntries(18000*1024*1024, bucketSize+avgEntrySize)
func BenchmarkEmptyMap(b *testing.B) {
       for i := 0; i < b.N; i++ {
               x = make(map[wire.OutPoint]*UtxoEntry, entries)
       }
}
```

Looks like the old code of `(18_000*1024*1024)/avgEntrySize` will make the map allocate N=26 and go right over the limit from  the get go.